apps/src/p5lab/spritelab/libraries/PoemBotLibrary.js

```diff
@@ -188,47 +179,167 @@ export default class PoemBotLibrary extends CoreLibrary {
       yCursor -= LINE_HEIGHT / 2;
-      this.p5.textSize(this.getScaledFontSize(this.poem.author, 16));
-      this.p5.text(this.poem.author, PLAYSPACE_SIZE / 2, yCursor);
-      this.p5.textSize(FONT_SIZE);
+      renderInfo.author = {
+        text: poemState.author,
+        x: PLAYSPACE_SIZE / 2,
+        y: yCursor,
+        size: this.getScaledFontSize(poemState.author, poemState.font, 16)
+      };
       yCursor += LINE_HEIGHT;
     }
-
-    const lineHeight = (PLAYSPACE_SIZE - yCursor) / this.poem.lines.length;
-    this.poem.lines.slice(0, this.currentLine).forEach(line => {
-      this.drawPoemLine(line, yCursor);
-      this.p5.textSize(FONT_SIZE);
+    const lineHeight = (PLAYSPACE_SIZE - yCursor) / poemState.lines.length;
+    poemState.lines.forEach(line => {
+      renderInfo.lines.push({
+        text: line,
+        x: PLAYSPACE_SIZE / 2,
+        y: yCursor,
+        size: this.getScaledFontSize(line, poemState.font, FONT_SIZE)
+      });
       yCursor += lineHeight;
     });
+    renderInfo = this.applyGlobalLineAnimation(renderInfo, frameCount);
+    poemState.effects.forEach(effect => {
+      renderInfo = this.applyEffect(renderInfo, effect, frameCount);
+    });
+    return renderInfo;
   }
 
-  drawPoemLine(line, yPos) {
-    // Concatenate all the text values together so we can compute the length
-    // of the printed text
-    const fullLine = line.map(textItem => textItem.value).join(' ');
-    this.p5.textSize(this.getScaledFontSize(fullLine, FONT_SIZE));
-
-    // compute line width with scaled textSize
-    let fullWidth = this.p5.textWidth(fullLine);
-
-    const start = PLAYSPACE_SIZE / 2 - fullWidth / 2;
-    this.p5.textAlign(this.p5.LEFT);
-    let xCursor = start;
-    line.forEach(textItem => {
-      this.p5.text(textItem.value, xCursor, yPos);
-      xCursor += this.p5.textWidth(textItem.value + ' ');
+  drawFromRenderInfo(renderInfo) {
+    this.p5.fill(renderInfo.color || 'black');
+    this.p5.textFont(renderInfo.font || 'Arial');
+    if (renderInfo.title) {
+      this.p5.textSize(renderInfo.title.size);
+      this.p5.text(
+        renderInfo.title.text,
+        renderInfo.title.x,
+        renderInfo.title.y
+      );
+    }
+    if (renderInfo.author) {
+      this.p5.textSize(renderInfo.author.size);
+      this.p5.text(
+        renderInfo.author.text,
+        renderInfo.author.x,
+        renderInfo.author.y
+      );
+    }
+    renderInfo.lines.forEach(item => {
+      let color = this.getP5Color(renderInfo.color || 'black', item.alpha);
+      this.p5.fill(color);
+      this.p5.textSize(item.size);
+      this.p5.text(item.text, item.x, item.y);
     });
   }
+
+  // polyfill for https://github.com/processing/p5.js/blob/main/src/color/p5.Color.js#L355
+  getP5Color(hex, alpha) {
+    let color = this.p5.color(hex);
+    if (alpha !== undefined) {
+      color._array[3] = alpha / color.maxes[color.mode][3];
+    }
+    const array = color._array;
+    // (loop backwards for performance)
```

Comment:
Interesting; what makes it faster?

Reply:
This is copied directly from the `p5.js` repo since we have a known bug with alpha values in our p5 fork (which I believe @epeach is going to fix before HOC launch, at which point we can take out this copied code). I found this PR though: https://github.com/processing/p5.js/pull/2831, which has some details on the performance optimizations.